### PR TITLE
Cjian/multi stage packaging pipeline

### PR DESCRIPTION
**Pull request by jchen351**
### Description

This will divide all jobs into stages.
This also divide MacOS packaging stages into 3 parallel job. This will effective reduce the build time by 3+ hours. 

### Motivation and Context

This way we can both run them in parallel also choose the one that we want to run instead of running every thing.

### CODE REVIEW

**Comment by snnn** [comment]
Looks like it is still failing.

File: tools/ci_build/github/azure-pipelines/templates/android-java-api-aar.yml [hunk 1]
```diff
@@ -40,69 +40,71 @@ parameters:
...
-          -e PACKAGE_NAME=${{parameters.packageName}} \
-          onnxruntimecpubuild \
-            /bin/bash /onnxruntime_src/tools/ci_build/github/android/build_aar_and_copy_artifacts.sh
-      workingDirectory: $(Build.SourcesDirectory)
-
-  - task: PublishBuildArtifacts@1
-    inputs:
-      pathtoPublish: '$(artifacts_directory)'
-      artifactName: '${{parameters.artifactName}}'
+stages:
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/android-java-api-aar.yml:43]: will need to update other usages of this template too, e.g.:
https://github.com/microsoft/onnxruntime/blob/bd142bfb04db00253265203fb0af2988e751d9a0/tools/ci_build/github/azure-pipelines/build-perf-test-binaries-pipeline.yml#L24
https://github.com/microsoft/onnxruntime/blob/bd142bfb04db00253265203fb0af2988e751d9a0/tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml#L24

File: tools/ci_build/github/azure-pipelines/templates/android-java-api-aar.yml [hunk 2]
```diff
@@ -40,69 +40,71 @@ parameters:
...
+- stage: Android_Java_API_AAR_Packaging_${{ parameters.job_name_suffix }}
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/android-java-api-aar.yml:44]: nit: "parameters.stage_name_suffix" would be more accurate now

File: tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml [hunk 3]
```diff
@@ -31,11 +31,11 @@ jobs:
     #  packageName: 'onnxruntime-mobile'
     buildSettings: '$(Build.SourcesDirectory)/tools/ci_build/github/js/react_native_e2e_full_aar_build_settings.json'
     artifactName: 'onnxruntime-android-full-aar'
-    job_name_suffix: 'For_React_Native'
+    stage_name_suffix: 'For_React_Native'
     pool_name: '${{parameters.PoolName}}'
     packageName: 'onnxruntime-android'
 
-- job: ReactNative_CI
+- stage: ReactNative_CI
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml:38]: I think it needs `jobs` before `steps`

File: tools/ci_build/github/azure-pipelines/nodejs/templates/test_linux.yml [hunk 4]
```diff
@@ -1,19 +1,21 @@
...
   JobSuffix: ''
-
-jobs:
-- job: Nodejs_Test_${{ parameters.JobSuffix }}
-  workspace:
-    clean: all
-  timeoutInMinutes:  120
-  pool: ${{ parameters.AgentPool }}
+stages:
+- stage: Nodejs_Test_${{ parameters.JobSuffix }}
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/nodejs/templates/test_linux.yml:5]: nit: also can rename "JobSuffix" to "StageSuffix", here and in other build definitions

File: tools/ci_build/github/azure-pipelines/nuget/templates/dml-vs-2019.yml [hunk 5]
```diff
@@ -22,252 +22,254 @@ parameters:
...
-    OnnxRuntimeBuildDirectory: '$(Build.BinariesDirectory)'
-    DotNetExe: 'dotnet.exe'
-    CUDA_VERSION: ${{ parameters.CudaVersion }}
-    runCodesignValidationInjection: and(${{ parameters.DoNodejsPack }},${{ parameters. DoEsrp}}) #For the others, code sign is in a separated job
-    DOTNET_SKIP_FIRST_TIME_EXPERIENCE: true
-    ALLOW_RELEASED_ONNX_OPSET_ONLY: ${{ parameters.AllowReleasedOpsetOnly }}
-    ${{ if eq(parameters.EnableLto, true) }}:
-      build_py_lto_flag: --enable_lto
+stages:
+- stage: ${{ parameters.JobName }}
```
Comment (edgchen1) [inline tools/ci_build/github/azure-pipelines/nuget/templates/dml-vs-2019.yml:26]: nit: rename "JobName" to "StageName"

**Comment by edgchen1** [comment]
looks like the pipeline "ONNX Runtime React Native CI Pipeline" is unable to be run:
/tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml (Line: 23, Col: 1): Unexpected value 'stages'

**Comment by jchen351** [comment]
> looks like the pipeline "ONNX Runtime React Native CI Pipeline" (https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=188) is unable to be run: /tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml (Line: 23, Col: 1): Unexpected value 'stages'



> 

Fixed by rolling back r[react-native-ci.yml](https://github.com/microsoft/onnxruntime/pull/14993/files/c80c46a7975439ce09218ae1eff9f07321acc59f#diff-8ad4338113e864dc9405ba4a407e87044dc936b90457e97ded859bcd63c7ba7f) and its associated files.

**Review by snnn — APPROVED** [review]

**Review by snnn — APPROVED** [review]
Nice job!

**Pull request merged**